internal/release/promote/pr_rendering.go

```diff
@@ -0,0 +1,200 @@
+package promote
+
+import (
+	"fmt"
+	"strings"
+	"text/template"
+
+	"github.com/Masterminds/sprig/v3"
+	"github.com/nestoca/joy/api/v1alpha1"
+	"golang.org/x/mod/semver"
+)
+
+type ChangeType string
+
+const (
+	ChangeTypeUpgrade   ChangeType = "Upgrade"
+	ChangeTypeDowngrade ChangeType = "Downgrade"
+	ChangeTypeUpdate    ChangeType = "Update"
+)
+
+type ReleaseInfo struct {
+	Name         string
+	Project      *v1alpha1.Project
+	CodeOwners   []string
+	Repository   string
+	Source       ReleaseWithGitTag
+	Target       ReleaseWithGitTag
+	OlderGitTag  string
+	NewerGitTag  string
+	IsPrerelease bool
+	// IsUpgrade is now deprecated and kept only temporarily while we transition to ChangeType
+	IsUpgrade  bool
+	ChangeType ChangeType
+	Commits    []*CommitInfo
+	Error      error
+}
+
+type PromotionInfo struct {
+	SourceEnvironment *v1alpha1.Environment
+	TargetEnvironment *v1alpha1.Environment
+	Releases          []*ReleaseInfo
+	Error             error
+}
+
+type CommitInfo struct {
+	Sha          string
+	ShortSha     string
+	Author       string
+	GitHubAuthor string
+	Message      string
+	ShortMessage string
+}
+
+func getReleaseInfo(sourceRelease, targetRelease *v1alpha1.Release, opts PerformOpts) *ReleaseInfo {
+	getAndPrintReleaseInfoWithError := func(msg string, args ...any) *ReleaseInfo {
```

Comment:
I think this function would make more send if it returned an error `getReleaseInfo(...) (*ReleaseInfo, error)`.
You are wrapping the error with information that is available outside the function, and it also should be the responsibility of the caller to handle the error:

```go
info, err := getReleaseInfo(source, target, opts)
if err != nil {
  fmt.Printf("⚠️ failed to get info for release %s: %v\n", source.Name, err)
  info = &ReleaseInfo{Name: source.Name, Error: err}
}
```

It's a great example of when you can handle an error by logging it, and continue on your way. Functions that generate errors shouldn't handle themselves.